overlord/snapstate/target.go

```diff
@@ -266,37 +294,92 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
 			snapst = &SnapState{}
 		}
 
-		// TODO: extract components from resources
-
 		// TODO: is it safe to pull the channel from here? i'm not sure what
 		// this will actually look like as a response from the real store
 		channel := r.RedirectChannel
 		if r.RedirectChannel == "" {
 			channel = sn.RevOpts.Channel
 		}
 
+		comps, err := requestedComponentsFromActionResult(sn, r)
+		if err != nil {
+			return nil, fmt.Errorf("cannot extract components from snap resources: %w", err)
+		}
+
 		installs = append(installs, target{
-			setup: &SnapSetup{
+			setup: SnapSetup{
 				DownloadInfo: &r.DownloadInfo,
 				Channel:      channel,
 				CohortKey:    sn.RevOpts.CohortKey,
 			},
 			info:       r.Info,
 			snapst:     *snapst,
-			components: nil,
+			components: comps,
 		})
 	}
 
 	return installs, err
 }
 
+func requestedComponentsFromActionResult(sn StoreSnap, sar store.SnapActionResult) ([]componentTarget, error) {
+	mapping := make(map[string]store.SnapResourceResult, len(sar.Resources))
+	for _, res := range sar.Resources {
+		mapping[res.Name] = res
+	}
+
+	installables := make([]componentTarget, 0, len(sn.Components))
+	for _, comp := range sn.Components {
+		res, ok := mapping[comp]
+		if !ok {
+			return nil, fmt.Errorf("cannot find component %q in snap resources", comp)
+		}
+
+		installable, err := componentFromResource(comp, res, sar.Info)
+		if err != nil {
+			return nil, err
+		}
+
+		installables = append(installables, installable)
+	}
+	return installables, nil
+}
+
+func componentFromResource(name string, sar store.SnapResourceResult, info *snap.Info) (componentTarget, error) {
+	comp, ok := info.Components[name]
+	if !ok {
+		return componentTarget{}, fmt.Errorf("%q is not a component for snap %q", name, info.SnapName())
+	}
+
+	if typ := fmt.Sprintf("component/%s", comp.Type); typ != sar.Type {
+		return componentTarget{}, fmt.Errorf("inconsistent component type (%q in snap, %q in component)", typ, sar.Type)
+	}
+
+	compName := naming.NewComponentRef(info.SnapName(), name)
+
+	return componentTarget{
+		setup: ComponentSetup{
+			DownloadInfo: &sar.DownloadInfo,
+		},
+		info: &snap.ComponentInfo{
+			Component: compName,
+			Type:      comp.Type,
+			Version:   sar.Version,
```

Comment:
Should we sanity chech the version?

Reply:
`snap.Component` doesn't carry a `Version`, is there something else you were thinking to compare it to?

Reply:
It would need to be read from component.yaml, I guess we can skip. Not sure though if what we put in `store.SnapActionResult` is checked against snap.yaml or not.